modules/lfs/http_client.go

```diff
@@ -134,71 +137,84 @@ func (c *HTTPClient) performOperation(ctx context.Context, objects []Pointer, dc
 		return fmt.Errorf("TransferAdapter not found: %s", result.Transfer)
 	}
 
+	errGroup, groupCtx := errgroup.WithContext(context.Background())
 	for _, object := range result.Objects {
-		if object.Error != nil {
-			log.Trace("Error on object %v: %v", object.Pointer, object.Error)
-			if uc != nil {
-				if _, err := uc(object.Pointer, object.Error); err != nil {
-					return err
-				}
-			} else {
-				if err := dc(object.Pointer, nil, object.Error); err != nil {
-					return err
-				}
-			}
-			continue
-		}
+		func(groupCtx context.Context, object *ObjectResponse, dc DownloadCallback, uc UploadCallback, transferAdapter TransferAdapter) {
+			errGroup.Go(func() error {
+				err := performSingleOperation(groupCtx, object, dc, uc, transferAdapter)
+				return err
+			})
+		}(groupCtx, object, dc, uc, transferAdapter)
+	}
 
+	// only the first error is returned, preserving legacy behavior before concurrency
+	return errGroup.Wait()
```

Comment:
https://pkg.go.dev/golang.org/x/sync/errgroup#Group.Wait

> Wait blocks until all function calls from the Go method have returned, then returns the first non-nil error (if any) from them. 

The operation should stop at first error, but not keep running when error occurs?

Reply:
It does though, when an error is returned from any of the threads it cancels all the other threads.

Reply:
Really?

```
func TestFoo(t *testing.T) {
	eg := errgroup.Group{}
	eg.SetLimit(2)
	c := atomic.Int64{}
	for i := 0; i < 10; i++ {
		eg.Go(func() error {
			time.Sleep(100 * time.Millisecond)
			c.Add(1)
			return io.ErrUnexpectedEOF
		})
	}
	assert.Error(t, eg.Wait())
	t.Logf("c: %d", c.Load())
	assert.NotEqualValues(t, 10, c.Load())
}
```


```
=== RUN   TestFoo
    pkg1_test.go:31: c: 10
    pkg1_test.go:32: 
        	Error Trace:	/Users/xiaoguang/work/test/playground-go/pkg1/pkg1_test.go:32
        	Error:      	Should not be: 10
        	Test:       	TestFoo
--- FAIL: TestFoo (0.50s)
```

Reply:
It's unclear to me what change I'm being asked to make. The function as it stands would attempt to download all the lfs objects in the batch in parallel. If any one of them returns an error, no further batches would be executed.

Are you perhaps asking me to cancel all the other (potentially successful) lfs object downloads within the same batch if one of them errored out early? I don't think that's the behavior we want, I think the behavior here is desired.

Reply:
> If any one of them returns an error, no further batches would be executed.

I agree. Maybe I misunderstood the `for loop`. I see now each single "download" / "upload" operation is a "batch", so the operation could stop when a batch fails.